providerquerymanager/providerquerymanager.go

```diff
@@ -0,0 +1,407 @@
+		return nil
+	case <-sessionCtx.Done():
+		return nil
+	}
+
+	var receivedInProgressRequest inProgressRequest
+	select {
+	case <-pqm.ctx.Done():
+		return nil
+	case <-sessionCtx.Done():
+		return nil
+	case receivedInProgressRequest = <-inProgressRequestChan:
+	}
+
+	return pqm.receiveProviders(sessionCtx, k, receivedInProgressRequest)
+}
+
+func (pqm *ProviderQueryManager) receiveProviders(sessionCtx context.Context, k cid.Cid, receivedInProgressRequest inProgressRequest) <-chan peer.ID {
+	// maintains an unbuffered queue for incoming providers for given request for a given session
+	// essentially, as a provider comes in, for a given CID, we want to immediately broadcast to all
+	// sessions that queried that CID, without worrying about whether the client code is actually
+	// reading from the returned channel -- so that the broadcast never blocks
+	// based on: https://medium.com/capital-one-tech/building-an-unbounded-channel-in-go-789e175cd2cd
+	returnedProviders := make(chan peer.ID)
+	receivedProviders := append([]peer.ID(nil), receivedInProgressRequest.providersSoFar[0:]...)
+	incomingProviders := receivedInProgressRequest.incoming
+
+	go func() {
+		defer close(returnedProviders)
+		outgoingProviders := func() chan<- peer.ID {
+			if len(receivedProviders) == 0 {
+				return nil
+			}
+			return returnedProviders
+		}
+		nextProvider := func() peer.ID {
+			if len(receivedProviders) == 0 {
+				return ""
+			}
+			return receivedProviders[0]
+		}
+		for len(receivedProviders) > 0 || incomingProviders != nil {
+			select {
+			case <-pqm.ctx.Done():
+				return
+			case <-sessionCtx.Done():
+				pqm.providerQueryMessages <- &cancelRequestMessage{
+					incomingProviders: incomingProviders,
+					k:                 k,
+				}
+				// clear out any remaining providers, in case and "incoming provider"
+				// messages get processed before our cancel message
+				for {
+					select {
+					case _, ok := <-incomingProviders:
+						if !ok {
+							return
+						}
+					case <-pqm.ctx.Done():
+						return
+					}
+				}
+			case provider, ok := <-incomingProviders:
+				if !ok {
+					incomingProviders = nil
+				} else {
+					receivedProviders = append(receivedProviders, provider)
+				}
+			case outgoingProviders() <- nextProvider():
+				receivedProviders = receivedProviders[1:]
+			}
+		}
+	}()
+	return returnedProviders
+}
+
+func (pqm *ProviderQueryManager) findProviderWorker() {
+	// findProviderWorker just cycles through incoming provider queries one
+	// at a time. We have six of these workers running at once
+	// to let requests go in parallel but keep them rate limited
+	for {
+		select {
+		case fpr, ok := <-pqm.providerRequestsProcessing:
+			if !ok {
+				return
+			}
+			k := fpr.k
+			log.Debugf("Beginning Find Provider Request for cid: %s", k.String())
+			pqm.timeoutMutex.RLock()
+			findProviderCtx, cancel := context.WithTimeout(fpr.ctx, pqm.findProviderTimeout)
+			pqm.timeoutMutex.RUnlock()
+			defer cancel()
```

Comment:
This will defer till the function returns.

Reply:
is that bad? this cancel now only serves to make sure the context gets cleaned up. the cancel due to all the requests cancelling is the parent of this context.

Reply:
Defered functions will get pushed onto a stack which will continue to grow until the worker returns.

Reply:
Basically, it'll leak a bunch of memory.

Reply:
Got it. I get a linter error if I don't name it and call it. I moved the discard to right after the for range of channel loop, which seems like the first safe time to cancel it.